### ISSUE TITLE
dask bag `take` ignores shared dependencies

### ISSUE THREAD

**Issue by alanhdu** [dask/dask#1438]
Normally, `take` seems to intelligently "propagate" its way down, so that a `take(1)` doesn't evaluate the entire task graph before . This is super useful for exploring large data sets.

Unfortunately, this doesn't seem to work for "diamond" / shared dependencies (like those created with `zip`), where the task graph looks like:

```
     --------
     | Take |
     --------
        /\
      /    \
    /        \
--------  --------
| Task |  | Task |
--------  --------
    \        /
      \    /
        \/
     --------
     | Base |
     --------
```

In this diamond dependency, even a trivial `take(1)` seems to evaluate the entire task graph, even though it should be easy to for `take(1)` to percolate down through the `zip`.

I saw this on Python 3.5 and dask `0.10.2` (both from `conda-forge`).

**Comment by alanhdu** [comment]
I've posted a notebook gist with a more concrete example:

https://gist.github.com/alanhdu/4545f306c06206ffa4d8cbe0a42ad2c6

Cells 5 and 6 are the important ones.

**Comment by alanhdu** [comment]
So on further investigation, this might just be a problem with `zip`. For some reason, `zip` seems _really_ slow. I have code that roughly looks like:

```python
import dask.bag as db
import json
bag = db.read_text("test*").map(json.loads)
ab = bag.pluck("a").pluck("b").map(some_computation)
c = bag.pluck("c")

with ProgressBar():
    # Computing them separately is pretty fast
    r1 = ab.compute()   # takes ~30 s
    r2 = c.compute()     # takes ~10 s
    results = list(zip(ab, c))

# Doing the zip with dask.bag is really slow
with ProgressBar():
    results = db.zip(ab, c).compute()    # 3 minutes later and still at 0% progress!
```

**Comment by mrocklin** [comment]
Yup, so here's what's going on.

Normally dask.bag partitions are iterators, they stream data if possible.  This means that it's even nicer than normal dask collections like dask.dataframe because we only need to compute a tiny bit of a partition to get something like take 1.  Because of this streaming-within-partitions ability dask.bag.read_text feels free to read each file as one partition by default.

This is great except when you tee off.  When this happens we no longer trust iterators (because each part of the tee would separately consume the same iterator) so we reify everything into a list.  Now we're still good about only computing the right partitions but we're bad about being lazy within each partition.

Unfortunately now our partition-per-file choice becomes bad, we're reading the entire file.  You can avoid this choice by explicitly sending in a `chunksize=10000000` or something, which will break the file up into 10MB blocks.  Then take(1) will only process 10000000 bytes instead of the entire file.  

Does this help?

**Comment by mrocklin** [comment]
Hrm, the results from the last zip example do look quite odd to me.

**Comment by alanhdu** [comment]
Explanation makes a lot of sense. Do you know if there's a way to tell dask to "fuse" operations together? Being able to operate on independent "columns" of the bag is a light easier than always `map` dictionaries. For example, it'd be nice if something like:
```Python
a = bag.pluck("a").map(something_complicated)
b = bag.pluck("b").map(something_else_complicated)
combined = db.zip(a, b)
```
did roughly the same thing as:
```Python
combined = bag.map(lambda x: {"a": something_complicated(x["a"]), 
                              "b": something_complicated(x["b"])})
```
(I realize they're not exactly the same because one returns a list of tuples while the other's a dictionary, but hopefully you get my point).

I find the first _much_ easier to read and maintain (especially if there are a bunch of chained `map` calls), and it'd be nice if it had the same performance characteristics regarding iterators.

**Comment by alanhdu** [comment]
Yeah, the zip is pretty confusing. A more end-to-end example goes like (see profile results http://nbviewer.jupyter.org/gist/alanhdu/6df4f7c64b7930acb29df0067f5bf975)
```Python
import json

from bokeh.io import output_notebook
from dask.diagnostics import Profiler, ProgressBar
import dask.bag as db

j = json.dumps({"a": 1, "b": 2})

data = [j for _ in range(4 * 10 ** 5)]
bag = db.from_sequence(data, npartitions=4).map(json.loads)

with Profiler() as prof1, ProgressBar():
    bag.pluck(["a", "b"]).count().compute()    # Takes 2.4 s

with Profiler() as prof2, ProgressBar():
    db.zip(bag.pluck("a"), bag.pluck("b")).count().compute()  # Takes 11.3 s
```
I can see possibly a 2x slowdown if the `zip` version has to redo all the json parsing (one for each pluck), but I'm not how it's a 5x slowdown (the 5x slowdown is pretty constant across sizes for `data` too). The profiling graphs look dask is optimizing the first version better, but I'm not really sure how to best interpret them.

**Comment by mrocklin** [comment]
One could write an optimization around zip that was run automatically.  It would be a function that consumes dask graphs, looks for cases of zip, and fuses them as you've one in your example.  If this is something you're interested in you could look at some of the other optimizations in `dask/optimize.py` and `dask/bag/core.py::optimize()` for examples.

**Comment by jakirkham** [comment]
It seems like it should still be possible (maybe even easier today) to optimize this case. Is there any interest in taking this on?
